Adding onboarding docs and configuring live preview (#5333)
ROCm/rocm-libraries: projects/hipdnn/docs/user-guides/how-to/get-set-engine-knob.rst

.. meta::
  :description: Learn how to get and set engine knob configurations in hipDNN.
  :keywords: hipDNN, ROCm, knobs

.. _set-engine-knobs:

********************************************
Get and set engine knob configurations in hipDNN
********************************************

> hipDNN has a flexible system of engine configuration knobs that lets plugin developers expose custom runtime settings that end users can then adjust.

Reviewer: ```suggestion
hipDNN has a flexible system of engine configuration knobs that lets hipDNN plugin developers expose custom runtime settings for end users to adjust.
```

Reviewer: This intro doesn't tell me what's going to happen in it

Author: Added

          std::cout << "Knob ID: " << knob.knobId() << "\n";
          std::cout << "Description: " << knob.description() << "\n";
          std::cout << "Type: " << static_cast<int>(knob.valueType()) << "\n";

          // Access default value (it's a variant)
          const auto& defaultVal = knob.defaultValue();
          if (std::holds_alternative<int64_t>(defaultVal)) {
              std::cout << "Default: " << std::get<int64_t>(defaultVal) << "\n";
          } else if (std::holds_alternative<double>(defaultVal)) {
              std::cout << "Default: " << std::get<double>(defaultVal) << "\n";
          } else if (std::holds_alternative<std::string>(defaultVal)) {
              std::cout << "Default: " << std::get<std::string>(defaultVal) << "\n";
          }

          // Check constraints
          const IConstraint* constraint = knob.constraint();
          if (constraint) {
              std::cout << "Constraint: " << constraint->toString() << "\n";
          }

          std::cout << "---\n";
      }
  } else {
      std::cerr << "Error getting knobs: " << error.get_message() << "\n";
  }

Use the knob lookup map
=======================

> Use the lookup method to access the knob ID:

Reviewer: Section title includes **map**. 
Intro includes **ID**.

Author: @CMiservaAMD: Should this say "Knob ID" or "Knob map"?

Reviewer 2: This question revealed a number of deficiences in this section. Content has been updated in https://github.com/ROCm/rocm-libraries/pull/5638

  KnobSetting stringSetting("test.knob", std::string("val")); // string

  // You can also update values later
  intSetting.setValue(100);

Use the default knob values
===========================

If you don't specify a knob setting, the engine will use the default value defined by the knob. To explicitly use defaults for all knobs:

.. code:: cpp

  // Option 1: Don't specify any settings (simplest)
  auto error = graph.create_execution_plan_ext(engineId, {});

  // Option 2: Specify only the knobs you want to customize
  std::vector<KnobSetting> settings;
  settings.emplace_back("global.benchmarking", 1);  // Only customize this one
  auto error = graph.create_execution_plan_ext(engineId, settings);

.. note::

  - All knob settings are validated against their constraints when creating an execution plan. Invalid values will result in an error with a descriptive message.
  - If you specify a knob that doesn't exist for the engine, hipDNN will log a warning but continue. This allows forward compatibility when new knobs are added.
  - If a knob is marked as deprecated, hipDNN will log a warning when you use it, but the knob will still function normally.

> Best practices

Reviewer 3: nested heading

>   - If a knob is marked as deprecated, hipDNN will log a warning when you use it, but the knob will still function normally.

Reviewer: ```suggestion
  - All knob settings are validated against their constraints when creating an execution plan. Invalid values result in an error with a descriptive message.
  - If you specify a knob that doesn't exist for the engine, hipDNN logs a warning and continues. This allows forward compatibility when new knobs are added.
  - If a knob is marked as deprecated, hipDNN logs a warning when you use it. The knob functions normally.
```

For production
--------------

- **Enable benchmarking during warm-up**:

  .. code:: cpp

    // Warm-up phase
    std::vector<KnobSetting> warmupSettings;
    warmupSettings.emplace_back("global.benchmarking", 1);
    graph.create_execution_plan_ext(engineId, warmupSettings);

    // Execute a few times to populate cache
    for (int i = 0; i < 5; i++) {
        graph.execute(handle, variantPack, workspace);
    }

- **Use cached results in production**: After warm-up, benchmarking can be disabled as results are cached.
- **Document knob settings**: Keep a record of knob configurations used in production for reproducibility.

For memory-constrained environments
-----------------------------------

- **Query workspace ranges**:

  .. code:: cpp

    // Find minimum and maximum workspace for the operation
    std::vector<Knob> knobs;
    graph.get_knobs_for_engine(engineId, knobs);

    for (const auto& knob : knobs) {
        if (knob.knobId() == "global.workspace_size_limit") {
            // Log constraint to understand valid range
        }
    }

- **Set conservative limits**: Start with a lower workspace limit and increase if performance is insufficient.

> - **Balance batch size and workspace**: Reducing workspace allows larger batch sizes, which may offset performance loss.

Reviewer 3: ```suggestion
- **Balance batch size and workspace**: Reducing workspace allows larger batch sizes, which might offset performance loss.
```
